### PR TITLE
Fix many things: less crashes, more stable

### PR DESCRIPTION
**Pull request by chase**
[View the changes that should be reviewed **_by everyone marked as a reviewer_**](https://github.com/coparse-inc/electron-libreoffice/pull/102/files?file-filters%5B%5D=.js&file-filters%5B%5D=.ts&show-viewed-files=true&show-deleted-files=false)

The value of return for reviewing the C++ is probably minimal, since we have a lot to do after this merges.

# Summary of Changes
## Better VS Code debugging!
- Install the CodeLLDB extension, go to the debugging panel, and attach to the PID of the renderer
- You can quickly get the PID in debug mode in the monorepo, or look at the logs from:
  - Chromium: `[684382:1124/170514.380926:ERROR:office_client.cc(224)] ...`, 684382 is the PID
  - LOK: `warn:i18nlangtag:684382:684468: ...`, 684382 is the PID
- After setting a breakpoint, you can fix the C++ standard library types to not be gibberish: go to the `Debug console` panel and run the `crfix` command that it mentions

## Simplification of the `libreoffice` global
- See the `OfficeClient` type in `libreoffice.d.ts`
- The event bus (`.on` `.off` `.emit`) and other unused APIs have been removed from the `libreoffice`
- All remaining APIs on `libreoffice` are now async, so they must be awaited in the code

## LOK no longer accessible in worker threads
- Since there is a global lock in an LOK process, there's not really a point to have it

## Object lifetime
- Threads can inherently outlive their callers, cases like this should no longer cause crashes
- Documents can held in a custom shared pointer type that's associated with a view, when nothing holds the document, it is released
- Reloading the page will release and close any opened documents
- Going to another page or closing the window will properly release and close any opened documents

## Renderer (`embed`)
- `renderDocument()` now accepts options
  - `zoom` option for initial zoom level
  - `disableInput` option to disable input
  - `restoreKey` option to restore a renderer's context and painting after it is removed from DOM
  - See `qa/index.js` and `qa/wrapper.js` changes for context and example

## Document client
- Creating new views no longer crashes
- `initializeForRendering()` is no longer a part of creating a DocumentClient, for documents that are intended to be rendered (not temporary) it must be called immediately after loading a document

## `qa/wrapper.js` and `qa/index.js`
- The `remount` functionality demonstrates similar behavior to what would occur in React if the `embed` is removed from the DOM then added back again, it uses the aforementioned `restoreKey` functionality

### CODE REVIEW

File: qa/index.js [hunk 1]
```diff
@@ -655,23 +533,16 @@ function colorizeWorker() {
...
-          });
-          xDoc.as('util.XCloseable').close(true);
+          doc.postUnoCommand('.uno:Colorize');
+          await doc.saveAs(data.file + '.pdf', 'pdf');
           self.postMessage({
             type: 'finished',
             time: (performance.now() - timeStart) / 1000,
           });
-          self.close();
+          // self.close();
```
Comment (seanaye) [inline qa/index.js:545]:
```suggestion
```

**Review by seanaye — APPROVED** [review]
Just reviewed the link diff but looks good. Easy to grok the js changes

**Review by synoet — APPROVED** [review]
Changes look good.

File: qa/index.js (hunk 1 above)
Comment (chase) [reply qa/index.js:545]: colorizeWorker itself should be removed since worker support has been removed as well, but I'll keep it there for reference

**Pull request merged**